scanpy/external/tl/_sam.py

```diff
@@ -0,0 +1,216 @@
+        `.layers['X_disp']`
+            The expression matrix used for nearest-neighbor averaging.
+        `.layers['X_knn_avg']`
+            The nearest-neighbor-averaged expression data used for computing the
+            spatial dispersions of genes.
+
+    Example
+    -------
+    >>> import scanpy.external as sce
+    >>> import scanpy as sc
+
+    *** Running SAM ***
+
+    Assuming we are given an AnnData object called `adata`, we can run the SAM
+    algorithm as follows:
+
+    >>> sam,adata = sce.tl.SAM(adata,inplace=True)
+
+    The input AnnData object should contain unstandardized, non-negative
+    expression values. Preferably, the data should be log-normalized and no
+    genes should be filtered out.
+
+    Please see the documentation for a description of all available parameters.
+
+    For more detailed tutorials, please visit the original Github repository:
+    https://github.com/atarashansky/self-assembling-manifold/tree/master/tutorial
+
+    *** Plotting ***
+
+    To visualize the output, we can use the built-in `scatter` function (this
+    assumes that `matplotlib` is installed.)
+
+    >>> sam.scatter(projection = 'X_umap')
+
+    `scatter` accepts all keyword arguments used in the
+    `matplotlib.pyplot.scatter` function. Please visit the plotting tutorials
+    for more information:
+
+    https://github.com/atarashansky/self-assembling-manifold/tree/master/tutorial/SAM_Plotting
+
+    *** SAMGUI ***
+
+    SAM comes with the SAMGUI module, a graphical-user interface written with
+    `Plotly` and `ipythonwidgets` for interactively exploring and annotating
+    the scRNAseq data and running SAM.
+
+    Dependencies can be installed with Anaconda by following the instructions in
+    the self-assembling-manifold Github README:
+    https://github.com/atarashansky/self-assembling-manifold
+
+    In a Jupyter notebook, execute the following to launch the interface:
+
+    >>> from SAMGUI import SAMGUI
+    >>> sam_gui = SAMGUI(sam) # sam is your SAM object
+    >>> sam_gui.SamPlot
+
+    This can also be enabled in Jupyer Lab by following the instructions in the
+    self-assembling-manifold README.
+
+    """
+
+    logg.info('Self-assembling manifold')
+
+    try:
+        from SAM import SAM
+    except ImportError:
+        raise ImportError(
+            '\nplease install sam-algorithm: \n\n'
+            '\tgit clone git://github.com/atarashansky/self-assembling-manifold.git\n'
+            '\tcd self-assembling-manifold\n'
+            '\tpip install .'
+        )
+
+    s = SAM(counts=adata, inplace=inplace)
+
+    logg.info('Running SAM')
+    s.run(
+        max_iter=max_iter,
+        num_norm_avg=num_norm_avg,
+        k=k,
+        distance=distance,
+        preprocessing=standardization,
+        weight_PCs=weight_pcs,
+        npcs=npcs,
+        n_genes=n_genes,
+        projection=projection,
+        verbose=verbose,
+    )
+
+    return (s, adata) if inplace else (s, s.adata)
```

Comment:
The return type (`Optional[AnnData]`) doesn’t match. I think the best would probably be a return type of `Union[SAM, AnnData]`:

```suggestion
    return s if inplace else s.adata
```

People expect `new_adata = sc.*tl.*(adata, inplace=False)` to work.

Reply:
Possibly a stupid question: `SAM` is defined within the scope of the function. If I want to set the return type to be `Union[SAM, AnnData]`, should I import SAM in the global frame (at the top of `_sam.py`)?

Reply:
Stupid of me, bumping the import statement to the global frame throws an ImportError since SAM is not a required package.

Reply:
Not a stupid question at all! That’s a complex thing we don’t have an easy solution for. What you can do is something like this:

```py
try:
    from sam import SAM
except ImportError:
    SAM = typing.Any
```